parser_library/src/workspaces/workspace_configuration.cpp

```diff
@@ -821,30 +843,165 @@ const processor_group& workspace_configuration::get_proc_grp(const proc_grp_id&
+
+    std::vector<diagnostic_s> diags;
+
+    auto proc_json = nlohmann::json::parse(group_json);
+    json_visitor(proc_json);
+    proc_json.get_to(pg);
+
+    if (!pg.asm_options.valid())
+        diags.push_back(diagnostic_s::error_W0005(normalized_location, pg.name, "external processor group"));
+    for (const auto& p : pg.preprocessors)
+    {
+        if (!p.valid())
+            diags.push_back(diagnostic_s::error_W0006(normalized_location, pg.name, p.type()));
+    }
+
+    processor_group prc_grp("", pg.asm_options, pg.preprocessors);
+
+    for (auto& lib_or_dataset : pg.libs)
+    {
+        std::visit(
+            [this, &diags, &prc_grp](const auto& lib) {
+                process_processor_group_library(lib, utils::resource::resource_location(), diags, {}, prc_grp);
+            },
+            lib_or_dataset);
+    }
+    m_utilized_settings_values.merge(std::move(utilized_settings_values));
+
+    for (auto&& d : diags)
+        prc_grp.add_diagnostic(std::move(d));
+
+    return m_proc_grps
+        .try_emplace(external_conf { std::make_shared<std::string>(std::move(group_json)) }, std::move(prc_grp))
+        .first;
+}
+
+void workspace_configuration::update_external_configuration(
+    const utils::resource::resource_location& normalized_location, std::string group_json)
+{
+    if (std::string_view group_name(group_json); utils::trim_left(group_name, " \t\n\r"), group_name.starts_with("\""))
+    {
+        m_exact_pgm_conf.insert_or_assign(normalized_location,
+            tagged_program {
+                program {
+                    normalized_location,
+                    basic_conf { nlohmann::json::parse(group_json).get<std::string>() },
+                    {},
+                    true,
+                },
+            });
+        return;
+    }
+
+    auto pg = m_proc_grps.find(tagged_string_view<external_conf> { group_json });
+    if (pg == m_proc_grps.end())
+    {
+        pg = make_external_proc_group(normalized_location, std::move(group_json));
+    }
+
+    m_exact_pgm_conf.insert_or_assign(normalized_location,
+        tagged_program {
+            program {
+                normalized_location,
+                pg->first,
+                {},
+                true,
+            },
+        });
+}
+
+
+void workspace_configuration::prune_external_processor_groups(const utils::resource::resource_location& location)
+{
+    if (!location.empty())
+    {
+        if (auto p = m_exact_pgm_conf.find(location.lexically_normal());
+            p != m_exact_pgm_conf.end() && p->second.pgm.external)
+            m_exact_pgm_conf.erase(p);
+    }
+    else
+        std::erase_if(m_exact_pgm_conf, [](const auto& p) { return p.second.pgm.external; });
+
+    std::erase_if(m_proc_grps, [](const auto& pg) {
+        const auto* e = std::get_if<external_conf>(&pg.first);
+        return e && e->definition.use_count() == 1;
```

Comment:
Maybe even:
```suggestion
        return e && e->definition.use_count() <= 1;
```
?

Reply:
`pg.first.definition` is the one reference we are testing for. `0` should not be possible.